test/r_tree/r_tree_test.dart

```diff
@@ -269,60 +269,127 @@ main() {
+/// bounding rectangles.
+_SubtreeValidationData assertNodeValidity<E>(RTree<E> tree, RTreeContributor contributor) {
   if (contributor is LeafNode<E>) {
-    if (contributor.height != 1) {
-      throw StateError('Leaf height of ${contributor.height} should be 1.');
-    }
-
-    return 1;
+    return assertLeafNodeValidity(tree, contributor);
   } else if (contributor is NonLeafNode<E>) {
-    var maxChildHeight = 0;
-    if (contributor.children.isNotEmpty) {
-      for (final child in contributor.children) {
-        final childHeight = assertNodeHeightValidity(tree, child);
-        if (childHeight > maxChildHeight) {
-          maxChildHeight = childHeight;
-        }
+    return assertNonLeafNodeValidity(tree, contributor);
+  }
+
+  // This is a datum
+  return _SubtreeValidationData(0, contributor.rect);
+}
+
+/// Comprehensively assert the consistency of the subtree rooted at the specified leaf node, including node height,
+/// parent references, and bounding rectangles.
+_SubtreeValidationData assertLeafNodeValidity<E>(RTree<E> tree, LeafNode<E> node) {
+  if (node.height != 1) {
+    throw StateError('Leaf height of ${node.height} should be 1.');
+  }
+
+  const defaultRect = Rectangle<num>(0, 0, 0, 0);
+  var actualRect = defaultRect;
+
+  if (node.children.isNotEmpty) {
+    for (final child in node.children) {
+      // Expand the node's rect to include this child
+      if (actualRect == defaultRect) {
+        actualRect = child.rect;
+      } else {
+        actualRect = actualRect.boundingBox(child.rect);
       }
     }
+  }
+
+  // Assert this node's rect/bounds match its actual structure
+  if (node.rect != actualRect) {
+    throw StateError('Leaf rect ${node.rect} should be $actualRect.');
+  }
 
-    final actualNodeHeight = 1 + maxChildHeight;
-    if (contributor.height != actualNodeHeight) {
-      throw StateError('Non-leaf height of ${contributor.height} should be $actualNodeHeight.');
+  return _SubtreeValidationData(1, actualRect);
+}
+
+/// Comprehensively assert the consistency of the subtree rooted at the specified non-leaf node, including node height,
+/// parent references, and bounding rectangles.
+_SubtreeValidationData assertNonLeafNodeValidity<E>(RTree<E> tree, NonLeafNode<E> node) {
+  const defaultRect = Rectangle<num>(0, 0, 0, 0);
+  var actualRect = defaultRect;
+  int? maxChildHeight;
+
+  if (node.children.isNotEmpty) {
+    for (final child in node.children) {
+      if (child.parent != node) {
+        throw StateError("Non-leaf child's parent reference is incorrect.");
+      }
+
+      // Traverse the tree from this child and collect validation data to propagate upwards
+      final childValidationData = assertNodeValidity(tree, child);
+
+      // Expand the node's rect to include this child
+      final childRect = childValidationData.rect;
+      if (actualRect == defaultRect) {
```

Comment:
this logic could also be shared (it would mean iterating the children again but I think that's okay in a test context)